### PR TITLE
protocol: reject duplicate connections

### PR DESCRIPTION
**Pull request by kim**
The idea to eject previous connections from a remote peer is racy, as described in #431. Instead, we reject new incoming connections until the old one is either closed gracefully, or times out.

### CODE REVIEW

**Review by FintanH — DISMISSED** [review]
LGTM :metal:

File: librad/src/net/protocol.rs [hunk 1]
```diff
@@ -503,18 +518,32 @@ where
...
+
+            conn.close(CloseReason::DuplicateConnection);
+            drop(incoming);
+
             self.subscribers
-                .emit(ProtocolEvent::Connected(remote_id))
+                .emit(ProtocolEvent::Disconnecting(remote_id))
                 .await;
-        }
+            drop(connections);
```
Comment (xla) [inline librad/src/net/protocol.rs:531]: drop

Comment (xla) [reply librad/src/net/protocol.rs:531]: What does this drop do? Is it to guarantee that the mutexguard is dropped?

**Review by xla — COMMENTED** [review]
I dunno if it is relevant in the context of this PR, but what does when the gossip stream terminates, shouldn't the connection be disconnected as well?  Currently it's only logged:

https://github.com/radicle-dev/radicle-link/blob/c65030894d51a7ec39fdbeb24d901ade5adb09b7/librad/src/net/protocol.rs#L528-L532

Alternatively is there a way to bring up a new gossip stream? The reason I'm bringing this up is that we observed the gossip stream exiting without the protocol bubbling up a disconnect event.

File: librad/src/net/protocol.rs (hunk 1 above)
Comment (kim) [reply librad/src/net/protocol.rs:531]: Yes

**Review by xla — APPROVED** [review]
🔱 🛀 🤑 🛤

**Comment by kim** [comment]
>  when the gossip stream terminates, shouldn't the connection be disconnected as well

Yeah right -- it shouldn't. The connection itself is actually independent of gossip membership. There is currently code which terminates the connection when _gossip_ decides to eject a peer, but that should be removed.

> Alternatively is there a way to bring up a new gossip stream

That's indeed a bit of a problem currently, because we keep the connection alive even if it is completely idle (ie. there is no `git` stream either). This means that it can only become active again if either end promotes it to the active set.

**Comment by xla** [comment]
Can we consider #431 fixed with this changeset?

**Review by FintanH — APPROVED** [review]

**Review by xla — APPROVED** [review]
📝 🍡 🗼 💁

**Review by xla — APPROVED** [review]
🌵 📣 🚇 🍛

**Comment by kim** [comment]
I suspect I might be missing a `spawn` somewhere...

**Comment by FintanH** [comment]
> I suspect I might be missing a spawn somewhere...

Here you go
image

**Review by FintanH — DISMISSED** [review]
Looks good to me

**Comment by kim** [comment]
That's a beautiful spawn.

Ya, so, I'm not sure. We only spawn incoming streams. Need to revisit that at some point.

**Comment by kim** [comment]
> is there a way to bring up a new gossip stream?

This is not properly addressed yet.

**Comment by xla** [comment]
@kim You wanna look at this before merging, or can go forward with this one?

**Comment by kim** [comment]
This should be approximately correct now (watch out for `XXX`). Is there are way we can test this in production?

**Comment by xla** [comment]
> This should be approximately correct now (watch out for `XXX`). Is there are way we can test this in production?

Yes, something specifically? Or just the quick reconnect of peers to the seed?

**Comment by kim** [comment]
> the quick reconnect of peers to the seed

This

**Comment by kim** [comment]
@xla does it seem alright?

**Comment by xla** [comment]
> @xla does it seem alright?

Testing as we speak.

**Review by xla — APPROVED** [review]
🚌 👚 🦁 🐹

**Review by FintanH — APPROVED** [review]

**Pull request merged**